### PR TITLE
Add option to use HSI as clock source for stm32 devices

### PR DESCRIPTION
**Pull request by fmanno**
- Add new compile flag MICROPY_HW_CLK_SRC_HSI=1
  - Modify documentation to include HSI

### CODE REVIEW

File: ports/stm32/Makefile [hunk 1]
```diff
@@ -347,6 +347,10 @@ SRC_USBDEV = $(addprefix $(USBDEV_DIR)/,\
 	class/src/usbd_msc_data.c \
 	)
 
+ifeq ($(MICROPY_HW_CLK_SRC_HSI),1)
+CFLAGS_MOD += -DMICROPY_HW_CLK_SRC_HSI=1
+endif
```
Comment (dpgeorge) [inline ports/stm32/Makefile]: It doesn't look like `MICROPY_HW_CLK_SRC_HSI` needs to be something you set via the Makefile (eg it doesn't include/exclude certain .c files in the build).  For options that don't need to be via the Makefile it's best to keep them just internal to C.  Basically these lines are not needed.

File: ports/stm32/boards/NUCLEO_F401RE/mpconfigboard.h [hunk 2]
```diff
@@ -5,8 +5,15 @@
...
-// HSE is 8MHz, CPU freq set to 84MHz
+// HSE is 8MHz, HSI is 16MHz CPU freq set to 84MHz
+// Default source for the clock is HSE, for revisions greater than C-01
+// Revision C-01 of the board isn't configured to use HSE
+// and must use HSI as a clock source
+#if MICROPY_HW_CLK_SRC_HSI
+#define MICROPY_HW_CLK_PLLM (16)
+#else
 #define MICROPY_HW_CLK_PLLM (8)
+#endif
```
Comment (dpgeorge) [inline ports/stm32/boards/NUCLEO_F401RE/mpconfigboard.h]: I think it's fair enough to have the default for this board as using HSI, as an example of how it works.  Then a comment saying how to change to HSE if needed.

File: ports/stm32/mpconfigboard_common.h [hunk 3]
```diff
@@ -181,11 +181,26 @@
 #error Unsupported MCU series
 #endif
 
-// Configure HSE for bypass or oscillator
-#if MICROPY_HW_CLK_USE_BYPASS
-#define MICROPY_HW_CLK_HSE_STATE (RCC_HSE_BYPASS)
-#else
-#define MICROPY_HW_CLK_HSE_STATE (RCC_HSE_ON)
+// Configure clock source for the chip
+#if MICROPY_HW_CLK_SRC_HSI
```
Comment (dpgeorge) [inline ports/stm32/mpconfigboard_common.h]: To be consistent with other config names, probably better to call this `MICROPY_HW_CLK_USE_HSI` instead (see eg `MICROPY_HW_RTC_USE_LSE`).

File: ports/stm32/mpconfigboard_common.h [hunk 4]
```diff
@@ -181,11 +181,26 @@
...
+#define MICROPY_RCC_OSCILLATOR_TYPE (RCC_OSCILLATORTYPE_HSI)
```
Comment (dpgeorge) [inline ports/stm32/mpconfigboard_common.h]: To keep the namespaces separated, please use `MICROPY_HW_RCC_xxx`.

File: ports/stm32/system_stm32.c [hunk 5]
```diff
@@ -538,7 +540,7 @@ void SystemClock_Config(void)
   }
 #endif
 
-    uint32_t vco_out = RCC_OscInitStruct.PLL.PLLN * (HSE_VALUE / 1000000) / RCC_OscInitStruct.PLL.PLLM;
+    uint32_t vco_out = RCC_OscInitStruct.PLL.PLLN * (HSI_VALUE / 1000000) / RCC_OscInitStruct.PLL.PLLM;
```
Comment (dpgeorge) [inline ports/stm32/system_stm32.c]: This is going to need a config or something because `HSE_VALUE` should be used if HSE is used.

**Review by dpgeorge — COMMENTED** [review]
Thanks for the contribution!  It's a good improvement.

**Comment by fmanno** [comment]
I've hopefully addressed your comments. I've also updated the wiki for this board

**Comment by dpgeorge** [comment]
Thanks for updating.  I made a few more minor edits (indenting and renamed existing MICROPY_HW_CLK_HSE_STATE to MICROPY_HW_RCC_HSE_STATE to make it consistent) and merged it in f938e70c6905474b08f7b5e6df3da9c1b0dcf01a

**Pull request closed without merging**